sound/soc/sof/ipc4.c

```diff
@@ -97,6 +97,7 @@ EXPORT_SYMBOL(sof_ipc4_process_reply);
 static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
 			     void *reply_data)
 {
+	struct sof_ipc4_msg *ipc4_msg = msg->reply_data;
```

Comment:
@ujfalusi I'm not able to appreciate why we need to do this. Maybe I'm missing the point. There's struct sof_ipc_msg that we use to pass the IPC message to be sent. What is the purpose of the sof_ipc4_msg? we dont have a sof_ipc3_msg similar to sof_ipc4_msg isnt it? So why is ipc4 special?

Reply:
@ranj063, all ipc3 messages start with `struct sof_ipc_cmd_hdr` followed by the payload, if any. When we receive the reply or notification then the header is behind of the msg->reply_data.
In IPC3 we pass the _whole_ message, including the header via mailbox.

IPC4 on the other hand is a 'split' protocol in this sense. The header is not passed via the mailbox, it is written to registers and the mailbox is only used in case there is additional data for the message.

We could do an asymmetric handling of IPC4, but I don't think it helps the reader by using the `struct sof_ipc4_msg` for TX and spreading out the reply, notification into `struct sof_ipc_msg`

Users of the IPC interface should not be using the `struct sof_ipc_msg` as it is internal to the IPC 'subsystem' of SOF, we should have a simple interface of "send this message and give me the reply here" and on the notification side it should be that, "here, we have received this message".
Ideally only the low level and high level should know how to deal with the pointer, in the middle layer we just pass things.
I consider the `ipc4_tx_wait_done()` here a low(ish) level functionality as it does error checking, but it should be possible to abstract it away to only wait w/o knowing anything about the content.

Reply:
Oh, I see now.
The RX for notification is handled completely differently than the reply in IPC3. The reply is copied to reply_data while the notification is not, the IPC3 RX function itself reads the message from the mailbox (first the header only, then the whole message when it knows the size).

Hrm, It might be better to rename the `reply_data` and `reply_size` to `rx_data` and `rx_size` and use it for both reply and notifications/events.
We are serializing things, so while a reply is handled it is guarantied that a notification will not change the `rx_data`.

That would make it a bit more well defined and we would save on `kmalloc()` for notifications as a bonus?